src/System.Device.Gpio/System/Device/Gpio/Drivers/SysFsDriver.Linux.cs

```diff
@@ -571,10 +641,56 @@ private unsafe void DetectEvents()
                     if (eventDetected)
                     {
-                        Thread.Sleep(1); // Adding some delay to make sure that the value of the File has been updated so that we will get the right event type.
-                        PinEventTypes eventTypes = (Read(pinNumber) == PinValue.High) ? PinEventTypes.Rising : PinEventTypes.Falling;
-                        var args = new PinValueChangedEventArgs(eventTypes, pinNumber);
-                        _devicePins[pinNumber]?.OnPinValueChanged(args);
+                        if (_statusUpdateSleepTime > 0)
+                        {
+                            Thread.Sleep(_statusUpdateSleepTime); // Adding some delay to make sure that the value of the File has been updated so that we will get the right event type.
+                        }
+
+                        var currentPin = _devicePins[pinNumber];
+                        var activeEdges = currentPin.ActiveEdges;
+                        PinEventTypes eventType = activeEdges;
+                        PinEventTypes secondEvent = PinEventTypes.None;
+                        // Only if the active edges are both, we need to query the current state and guess about the change
+                        if (activeEdges == (PinEventTypes.Falling | PinEventTypes.Rising))
+                        {
+                            var oldValue = currentPin.LastValue;
+                            var newValue = Read(pinNumber);
+                            if (oldValue == PinValue.Low && newValue == PinValue.High)
+                            {
+                                eventType = PinEventTypes.Rising;
+                            }
+                            else if (oldValue == PinValue.High && newValue == PinValue.Low)
+                            {
+                                eventType = PinEventTypes.Falling;
+                            }
+                            else if (oldValue == PinValue.High)
+                            {
+                                // Both high -> There must have been a low-active peak
+                                eventType = PinEventTypes.Falling;
+                                secondEvent = PinEventTypes.Rising;
+                            }
+                            else
+                            {
+                                // Both low -> There must have been a high-active peak
+                                eventType = PinEventTypes.Rising;
+                                secondEvent = PinEventTypes.Falling;
+                            }
+
+                            currentPin.LastValue = newValue;
+                        }
+                        else
+                        {
+                            // Update the value, in case we need it later
+                            currentPin.LastValue = Read(pinNumber);
```

Comment:
I think that this READ should be the very first thing we do after the thread.sleep above, since we want to avoid as much as possible some other events to happen while doing other calculations and this read operation to get wrong value.

Reply:
I agree. Changed.